### ISSUE TITLE
Project Rewrite

### ISSUE THREAD

**Issue by vyfor** [vyfor/cord.nvim#133]
This issue is used to track #131 (see PR for the implementation details).
It is a complete rewrite of the existing codebase and will introduce significant breaking changes.

**Comment by vyfor** [comment]
For now, all connected clients will share the same configuration and application, and no client will be able to override this until the server is shut down. We'll likely need to build a session manager to handle things that are unique to each session, such as workspace settings and git buttons, which are less prone to change across the session. Instead of the pipe server managing connected clients, the session manager will take over that responsibility. Once this is in place, we can start thinking about handling different configurations and, of course, creating separate rich client instances for each unique application ID, since Discord does not allow us to dynamically change the rich presence name (classic Discord moment), and on top of that, they impose heavy rate limits on clearing the rich presence and reconnecting to the pipe (classic Discord moment).

**Comment by vyfor** [comment]
I plan to introduce a number of new features into the existing configuration, as well as modify some existing fields:
```lua
usercmds = true,
log_level = vim.log.levels.ERROR, -- replace string values with built-in constants
timestamp = {
  enable = true,
  reset_on_idle = false,
  reset_on_change = false,
},
editor = {
  image = nil,
  client = 'neovim',
  tooltip = 'The Superior Text Editor',
},
display = {
  swap_fields = false,
  swap_icons = false,
},
lsp = {
  severity = 1,
  scope = 'workspace',
},
idle = {
  enable = true,
  timeout = 300000,
  show_status = true,
  ignore_focus = true, -- previously `disable_on_focus`
  icon = 'idle',
  text = 'Idling',
  tooltip = 'Zzz',
},
text = { -- now supports functions
  viewing = 'Viewing {}',
  editing = 'Editing {}',
  file_browser = 'Browsing files in {}',
  plugin_manager = 'Managing plugins in {}',
  lsp_manager = 'Configuring LSP in {}',
  docs = 'Viewing {} documentation',
  vcs = 'Committing changes in {}',
  dashboard = 'Home',
  workspace = function(opts)
    return 'Working in ' .. opts.name .. ' - ' .. opts.problem_count .. ' problems'
  end
},
buttons = {
  {
    label = 'View Repository',
    url = 'git', -- now supports functions
  },
},
assets = nil,
hooks = {
  on_ready = function() end,
  on_connect = function() end,
  on_disconnect = function() end,
  on_update = function() end,
  on_workspace_change = function() end,
}
```
With these changes, configuration options that previously relied on static values and are achievable with use of the new functions and hooks (fields like cursor position, problem count, and workspace blacklist) will be removed. This change allows for greater customization and will be documented on a wiki page, where examples and best practices on how to achieve various configurations using this new approach will be shown.

**Comment by vyfor** [comment]
The PR is nearly finished, with just a few final touches remaining. You can try it out with this lazy.nvim configuration, though bugs are expected. I'd appreciate any feedback you have.
```lua
{
  'vyfor/cord.nvim',
  branch = 'client-server',
  build = 'cargo build --release',
  opts = {},
}
```

**Comment by vyfor** [comment]
It might be a good idea to move the file out of the plugin's root directory (`target/release`). Keeping it there could make it harder for users to update or remove the plugin through plugins managers without first disabling it or stopping the server process.

**Comment by vyfor** [comment]
We're doing quite a few blocking I/O calls when spawning the server. We likely want to replace them with luvit's async api

**Comment by vyfor** [comment]
Cord now features **icon themes**! Users are able to switch between different themes that match their own style. For now, two themes are available: `onyx`, a sleek dark theme that is the new default, and `pastel`, a colorful theme which resembles Cord’s previous style. Many of the existing icons have been redesigned as well.
Check out the icon showcase here: https://github.com/vyfor/icons#showcase

**Comment by vyfor** [comment]
The rewrite is complete and ready for merging. Given the major breaking changes and limited testing, the merge is scheduled for January 1, 2025. However, this date may change depending on circumstances and whether the rewrite is stable. I expect any bugs, if any, to be resolved by then. The previous version of the plugin will still be available under the `v1.0.0` tag. If you encounter any unexpected behavior or bugs, please feel free to create a new issue. I would also greatly appreciate any feedback you may have.

P.S. Check out the updated [README](https://github.com/vyfor/cord.nvim/blob/client-server)!

**Comment by vyfor** [comment]
I'd like to implement proper versioning for the plugin. We cannot obviously follow the semver spec since we would have to create a release for the server even when there are no changes to the Rust codebase. And not doing so would make server version inconsistencies look bad. The plan is to bump the PATCH version for updates to the Lua codebase as they are frequent, the MINOR version for either:
A. Rust codebase updates, which I expect to be rare
B. Breaking changes
and the MAJOR version for major updates.
